README.md
Serf Cookbook
=============

[![Cookbook Version](https://img.shields.io/cookbook/v/serf.svg)](https://community.opscode.com/cookbooks/serf)

Installs and configures [Serf](http://www.serfdom.io/).

View the [Change Log](https://github.com/bbaugher/serf/blob/master/CHANGELOG.md) to see what has changed.

Supports
--------

 * Serf >= 0.3 (uses config reloading via SIGHUP)
 * Linux OS'es

Usage
-----

Using the default attributes will setup a single Serf agent in its own cluster.

If you already have a Serf agent (or cluster) running specify the array of address(es) with the
`node["serf"]["agent"]["start_join"]` attribute so the agent will join the cluster(s).

What does the installation look like
------------------------------------

By default the installation will look like,

    serf | /usr/bin/serf  - The serf binary command
    /opt/serf/*           - All of serf's files (config, binaries, event handlers, logs...)
    /etc/serf/*           - Link to all of serf's config files
    /var/log/serf/*       - Link to all of serf's log files
    /etc/init.d/serf      - An init.d script to start/stop the agent. You can use service
                    serf [start|stop|restart|status] instead

Event Handlers
--------------

An [event handler](http://www.serfdom.io/docs/agent/event-handlers.html) is a script that is run when the Serf agent
recieves an event (member-join, member-leave, member-failed, or user).

You can configure an event handler via the attribute `node["serf"]["event_handlers"]`. The format of the `event_handlers`
attribute is the following,

    [
      {
        "url" : "URL", # REQUIRED
        "event_type" : "EVENT_TYPE" #OPTIONAL
      },
      ...
    ]

Chef will download the event handler and ensure it stays up to date. It will also add it to the serf agent's list of event handlers.
Each event handler must have a unique name.

The `event_type` value filters the event handler for certain events. Use [this doc](http://www.serfdom.io/docs/agent/event-handlers.html)
to figure out the `event_type` you need.

It is also possible to add event handlers via the attribute `node["serf"]["agent"]["event_handlers"]`. The `node["serf"]["event_handlers"]`
helps with the deployment of the event handler file itself and will add the event handler to the `node["serf"]["agent"]["event_handlers"]`
attribute.

Resources and Providers
-----------------------
### `serf_event`
The `serf_event` LWRP resource dispatches a custom user event into a Serf cluster by executing serf CLI.

#### Actions
- :create: Executes serf cli with parameters required to send an event.

#### Attribute Parameters
- event_name: name attribute. The name of the event to send
- payload: Optional event's payload. Any string you want.
- coalesce: Coalesce option. `true` by default
- rpc_addr: Address of serf node to connect for event dispatching. Local node by default.
- rpc_auth: RPC auth token. If not set, `node['serf']['agent']['rpc_auth']` will be used.

#### Example

Send custom user event

```ruby
serf_event 'deploy' do
  payload '1234567890'
  rpc_auth 'secret'
end
```

It would look like this in serf log:

```
2014/07/08 22:53:07 [INFO] agent: Received event: user-event: deploy
Event Info:
    Coalesce: true
    Event: "user"
    LTime: 7
    Name: "deploy"
    Payload: []byte{0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30}
```

Attributes
----------

 * `node["serf"]["user"]` : The user that owns the serf installation (default="serf")
 * `node["serf"]["group"]` : The group that owns the serf installation (default="serf")
 * `node["serf"]["agent"][*]` : A hash of key/values that will be added to the agent's config file (default={}). Use [this doc](http://www.serfdom.io/docs/agent/options.html) to configure the agent.
 * `node["serf"]["event_handlers"]` : An array of hashes that represent [event handlers](http://www.serfdom.io/docs/agent/event-handlers.html). See 'Event Handlers' above for more details (default=[])
 * `node["serf"]["base_binary_url"]` : The base url used to download the binary zip (default="https://dl.bintray.com/mitchellh/serf/")
 * `node["serf"]["version"]` : The version of the Serf agent to install (default="0.3.0")
 * `node["serf"]["arch"]` : The architecture of the Serf agent to install (default=`kernel['machine'] =~ /x86_64/ ? "amd64" : "386"`)
 * `node["serf"]["binary_url"]` : The full binary url of the Serf agent. If you override this value make sure to provide a valid and up to date value for `node["serf"]["version"]` (default=`File.join node["serf"]["base_binary_url"], "#{node["serf"]["version"]}_linux_#{node["serf"]["arch"]}.zip"`)
 * `node["serf"]["base_directory"]` : The base directory Serf should be installed into (default="/opt/serf")
 * `node["serf"]["log_directory"]` : The directory of the Serf agent logs (default="/var/log/serf")
 * `node["serf"]["conf_directory"]` : The directory of the Serf agent config (default="/etc/serf")
 * `node["serf"]["on_config_change"]` : An action to be performed if config is changed (default=:reload)
 * `node["serf"]["init_info"]` : A hash of init information used by the init service script (see attributes file for defaults)

CHANGELOG.md
Change Log
==========

[1.2.0 - 08-16-2017](https://github.com/bbaugher/serf/issues?milestone=16&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-36](https://github.com/bbaugher/serf/issues/36) : Kernel Attribute Fix

[1.1.0 - 04-06-2016](https://github.com/bbaugher/serf/issues?milestone=15&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-33](https://github.com/bbaugher/serf/issues/33) : Initialize SerfHelper Differently to Avoid compat_resource Cookbook Conflict

[1.0.0 - 02-09-2016](https://github.com/bbaugher/serf/issues?milestone=14&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-32](https://github.com/bbaugher/serf/issues/32) : Updates serf download url to official repo

[0.9.0 - 04-23-2015](https://github.com/bbaugher/serf/issues?milestone=13&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-31](https://github.com/bbaugher/serf/issues/31) : Refactoring

[0.8.0 - 02-10-2015](https://github.com/bbaugher/serf/issues?milestone=12&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-30](https://github.com/bbaugher/serf/issues/30) : serf doesn't start on CentOS 7 with dhcp
  * [Enhancement] [Issue-10](https://github.com/bbaugher/serf/issues/10) : Add tests

[0.7.1 - 08-21-2014](https://github.com/bbaugher/serf/issues?milestone=11&state=closed)
---------------------------------------------------------------------------------------

  * [Bug] [Issue-29](https://github.com/bbaugher/serf/issues/29) : PaxHeader folder in 0.7.0 release

[0.7.0 - 07-30-2014](https://github.com/bbaugher/serf/issues?milestone=10&state=closed)
---------------------------------------------------------------------------------------

  * [Enhancement] [Issue-28](https://github.com/bbaugher/serf/issues/28) : Get action to perform on config change from attributes

[0.6.0 - 07-08-2014](https://github.com/bbaugher/serf/issues?milestone=9&state=closed)
--------------------------------------------------------------------------------------

  * [Feature] [Issue-27](https://github.com/bbaugher/serf/issues/27) : Feature/event

[0.5.0 - 03-30-2014](https://github.com/bbaugher/serf/issues?milestone=7&state=closed)
--------------------------------------------------------------------------------------

  * [Enhancement] [Issue-23](https://github.com/bbaugher/serf/issues/23) : Adding support for chkconfig.
  * [Enhancement] [Issue-22](https://github.com/bbaugher/serf/issues/22) : Version normalization
  * [Enhancement] [Issue-19](https://github.com/bbaugher/serf/issues/19) : Prettify serf agent json (humans too read configuration files).

[0.4.2 - 02-15-2014](https://github.com/bbaugher/serf/issues?milestone=8&state=closed)
--------------------------------------------------------------------------------------

  * [Enhancement] [Issue-21](https://github.com/bbaugher/serf/issues/21) : License?
  * [Bug] [Issue-20](https://github.com/bbaugher/serf/issues/20) : Event handlers attribute ignores type

[0.4.1 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=6&state=closed)
--------------------------------------------------------------------------------------

  * [Bug] [Issue-18](https://github.com/bbaugher/serf/issues/18) : Reload serf agent with "kill -HUP" instead of broken "kill -SIGHUP".
  * [Bug] [Issue-17](https://github.com/bbaugher/serf/issues/17) : Fixed logging of serf agent messages when /bin/sh is not a symlinked bash.

[0.4.0 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=3&state=closed)
--------------------------------------------------------------------------------------

  * [Feature] [Issue-12](https://github.com/bbaugher/serf/issues/12) : Provide ability to reload configuration without restart

[0.3.0 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=2&state=closed)
--------------------------------------------------------------------------------------

  * [Bug] [Issue-14](https://github.com/bbaugher/serf/issues/14) : Init script broken on ubuntu (regression)
  * [Enhancement] [Issue-11](https://github.com/bbaugher/serf/issues/11) : Create serf user that owns installation

[0.2.1 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=4&state=closed)
--------------------------------------------------------------------------------------

  * [Bug] [Issue-15](https://github.com/bbaugher/serf/issues/15) : /var/log/serf and /etc/serf link to files

[0.2.0 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=1&state=closed)
--------------------------------------------------------------------------------------

  * [Bug] [Issue-7](https://github.com/bbaugher/serf/issues/7) : Repair working init script on ubuntu :)
  * [Bug] [Issue-6](https://github.com/bbaugher/serf/issues/6) : fix init script to allow multiple join addresses
  * [Enhancement] [Issue-5](https://github.com/bbaugher/serf/issues/5) : Use kill -INT [PID] to gracefully shutdown agent
  * [Enhancement] [Issue-4](https://github.com/bbaugher/serf/issues/4) : Update README with information about where things live, services, logs
  * [Enhancement] [Issue-1](https://github.com/bbaugher/serf/issues/1) : Move configuration out of init.d to file

[0.1.0 - 01-22-2014](https://github.com/bbaugher/serf/issues?milestone=5&state=closed)
--------------------------------------------------------------------------------------

  * [Feature] [Issue-16](https://github.com/bbaugher/serf/issues/16) : Initial cookbook


test/integration/default/serverspec/default_spec.rb
# coding: UTF-8

require 'spec_helper'

describe user('serf') do
  it { should exist }
  it { should belong_to_group 'serf' }
end

describe service('serf') do
  it { should be_running   }
end

describe file('/etc/serf/serf_agent.json') do
  it { should be_file }
  it { should be_owned_by 'serf' }
  it { should be_grouped_into 'serf' }
  it { should contain '"log_level": "info"' }
end

describe file('/etc/init.d/serf') do
  it { should be_file }
  it { should be_owned_by 'serf' }
  it { should be_grouped_into 'serf' }
  it { should contain 'Provides: serf' }
  it { should contain 'Default-Start: 3 4 5' }
  it { should contain 'Default-Stop: 0 1 2 6' }
  it { should contain 'fake: value' }
  it { should contain 'chkconfig: 2345 95 20' }
end


providers/agent.rb
require 'json'

# Support whyrun
def whyrun_supported?
    true
end

action :create do
    if @current_resource.exists
        Chef::Log.info "#{ @new_resource } already exists - nothing to do."
    else
        converge_by("Create #{ @new_resource }") do
            create_serf_agent
        end
    end
end

action :start do
    if @current_resource.exists
        converge_by("Start #{ @new_resource }") do
            start_serf_agent
        end
    else
        Chef::Log.info "#{ @current_resource } doesn't exist - can't start."
    end
end

action :restart do
    if @current_resource.exists
        converge_by("Restart #{ @new_resource }") do
            restart_serf_agent
        end
    else
        Chef::Log.info "#{ @current_resource } doesn't exist - can't restart."
    end
end

action :stop do
    if @current_resource.exists
        converge_by("Stop #{ @new_resource }") do
            stop_serf_agent
        end
    else
        Chef::Log.info "#{ @current_resource } doesn't exist - can't stop."
    end
end

action :delete do
    if @current_resource.exists
        converge_by("Delete #{ @new_resource }") do
            delete_serf_agent
        end
    else
        Chef::Log.info "#{ @current_resource } doesn't exist - can't delete."
    end
end

def load_current_resource
    @current_resource = Chef::Resource::SerfAgent.new(@new_resource.name)
    @current_resource.name(@new_resource.name)
    @current_resource.group(@new_resource.group)
    @current_resource.user(@new_resource.user)
    @current_resource.base_directory(@new_resource.base_directory)
    @current_resource.log_directory(@new_resource.log_directory)
    @current_resource.conf_directory(@new_resource.conf_directory)
    @current_resource.agent(@new_resource.agent)
  
    if agent_exists?
        @current_resource.exists = true
    end

    @current_resource
end

private
    def agent_exists?
        return false unless ::File.exist?(::File.join(current_resource.conf_directory, "#{current_resource.name}.json") )
        return false unless ::File.exist?(::File.join("/etc/init.d", "#{current_resource.name}") )
        return false unless ::File.exist?(::File.join("/etc/logrotate.d/", "serf_agent_#{current_resource.name}") )
        current_agent = JSON.parse(::File.read(::File.join(current_resource.conf_directory, "#{current_resource.name}.json") ));
        new_resource.agent == current_agent
    end

    def create_serf_agent
        if new_resource.name != "serf"
            link ::File.join(new_resource.base_directory, "bin", new_resource.name) do
              to ::File.join(new_resource.base_directory, "bin", "serf")
            end
        end

        template "/etc/logrotate.d/serf_agent_#{new_resource.name}" do
            source  "serf_log_roll.erb"
            group new_resource.group
            owner new_resource.user
            mode 00755
            variables(:agent_log_file => ::File.join(current_resource.log_directory, "#{current_resource.name}.log"))
            backup false
        end

        service new_resource.name do
            supports :status => true, :restart => true, :reload => true
            action :nothing
        end

        template ::File.join(new_resource.conf_directory, "#{new_resource.name}.json") do
            source  "serf_agent.json.erb"
            group new_resource.group
            owner new_resource.user
            mode 00755
            variables( :agent_json => JSON.pretty_generate(new_resource.agent.to_hash.reject {|key, value| key == "name"}) )
            backup false
            notifies :restart, "service[#{new_resource.name}]"
        end

        template "/etc/init.d/#{new_resource.name}" do
            source  "serf_service.erb"
            group new_resource.group
            owner new_resource.user
            mode  00755
            variables(
                :serf_name => new_resource.name, 
                :serf_binary => ::File.join(new_resource.base_directory, "bin", new_resource.name), 
                :serf_log => ::File.join(new_resource.log_directory, "#{current_resource.name}.log"),
                :serf_conf => ::File.join(new_resource.conf_directory, "#{new_resource.name}.json"),
                :serf_user => new_resource.user)
            backup false
            notifies :restart, "service[#{new_resource.name}]"
        end

        # Ensure everything is owned by serf user/group
        execute "chown -R #{new_resource.user}:#{new_resource.group} #{new_resource.base_directory}"
    end

    def stop_serf_agent
        service new_resource.name do
            action :stop
        end
    end

    def start_serf_agent
        service new_resource.name do
            supports :status => true, :restart => true, :reload => true
            action [ :enable, :start ]
        end
    end

    def restart_serf_agent
        service new_resource.name do
            supports :status => true, :restart => true, :reload => true
            action [ :stop, :reload, :start ]
        end
    end

    def delete_serf_agent
        stop_serf_agent

        link ::File.join(current_resource.base_directory, "bin", current_resource.name) do
          action :delete
        end

        file "/etc/logrotate.d/serf_agent_#{current_resource.name}" do
          action :delete
        end

        file ::File.join(current_resource.conf_directory, "#{current_resource.name}.json") do
          action :delete
        end

        file ::File.join(current_resource.conf_directory, "/etc/init.d/#{current_resource.name}") do
          action :delete
        end
    end

libraries/serf_helper.rb
# coding: UTF-8
require 'json'

class Chef::Recipe::SerfHelper

  SERF_VERSION_REGEX = /^Serf v\d.\d.\d/
  VERSION_REGEX = /\d.\d.\d/

  attr_accessor :node

  def initialize(node)
    @node = node
  end

  def get_bin_directory
     File.join node["serf"]["base_directory"], "bin"
  end

  def get_serf_binary
    File.join get_bin_directory, "serf"
  end

  def get_event_handlers_directory
    File.join node["serf"]["base_directory"], "event_handlers"
  end

  def get_home_config_directory
    File.join node["serf"]["base_directory"], "config"
  end

  def get_agent_config
    File.join get_home_config_directory, "serf_agent.json"
  end

  def get_home_log_directory
    File.join node["serf"]["base_directory"], "logs"
  end

  def get_agent_log
    File.join get_home_log_directory, "agent.log"
  end

  def get_agent_json
    JSON.pretty_generate(node["serf"]["agent"].to_hash)
  end

  def get_zip_file_path
    File.join Chef::Config[:file_cache_path], get_binary_filename
  end

  def get_binary_url
    node["serf"]["binary_url"] ||
    File.join(node["serf"]["base_binary_url"], node["serf"]["version"], get_binary_filename)
  end

  def get_serf_installed_version
    unless File.exists? get_serf_binary
      return "NONE"
    end

    version_output = `#{get_serf_binary} version`.chomp
    serf_match = SERF_VERSION_REGEX.match(version_output)

    if serf_match.size == 0
      raise "Unable to parse version from `serf version` output [#{version_output}]"
    end

    version_match = VERSION_REGEX.match(serf_match[0])

    # Should never happen
    if version_match.size == 0
      raise "Unable to parse version from `serf version` output [#{version_output}]"
    end

    version_match[0]
  end

  private

  def get_binary_filename
    "serf_#{node["serf"]["version"]}_linux_#{node["serf"]["arch"]}.zip"
  end
end


attributes/default.rb
# coding: UTF-8

default["serf"]["user"] = "serf"
default["serf"]["group"] = "serf"

default["serf"]["on_config_change"] = :reload

default["serf"]["agent"] = {}
default["serf"]["agent"]["event_handlers"] = []

default["serf"]["event_handlers"] = []

default["serf"]["version"] = "0.7.0"
default['serf']['arch'] = node['kernel']['machine'] =~ /x86_64/ ? "amd64" : "386"

default["serf"]["base_binary_url"] = "https://releases.hashicorp.com/serf/"
default["serf"]["binary_url"] = nil

default["serf"]["base_directory"] = "/opt/serf"
default["serf"]["log_directory"] = "/var/log/serf"
default["serf"]["conf_directory"] = "/etc/serf"

default["serf"]["init_info"]["Provides"] = "serf"
default["serf"]["init_info"]["Short-Description"] = "Serf agent"
default["serf"]["init_info"]["Default-Start"] = "3 4 5"
default["serf"]["init_info"]["Default-Stop"] = "0 1 2 6"
default["serf"]["init_info"]["Required-Start"] = ""
default["serf"]["init_info"]["Required-Stop"] = ""
default["serf"]["init_info"]["Should-Start"] = ""
default["serf"]["init_info"]["Should-Stop"] = ""
default["serf"]["init_info"]["chkconfig"] = "2345 95 20"
default["serf"]["init_info"]["description"] = "Serf agent"
default["serf"]["init_info"]["processname"] = "serf"
